pkg/filesystem/nydus/resolver.go

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2021. Alibaba Cloud. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package nydus
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/containerd/containerd/reference/docker"
+	"github.com/containerd/nydus-snapshotter/pkg/auth"
+	"github.com/containerd/nydus-snapshotter/pkg/utils/registry"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/pkg/errors"
+)
+
+type Resolver struct {
+	transport http.RoundTripper
+}
+
+func NewResolver() *Resolver {
+	resolver := Resolver{
+		transport: http.DefaultTransport,
+	}
+	return &resolver
+}
+
+func (r *Resolver) Resolve(ref, digest string, labels map[string]string) (io.ReadCloser, error) {
+	named, err := docker.ParseDockerRef(ref)
+	if err != nil {
+		return nil, err
+	}
+	host := docker.Domain(named)
+	sref := fmt.Sprintf("%s/%s", host, docker.Path(named))
+	nref, err := name.ParseReference(sref)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse ref %q (%q)", sref, digest)
+	}
+	keychain := auth.GetRegistryKeyChain(host, labels)
+
+	var tr http.RoundTripper
+	if nref.Context().Registry.Scheme() == "https" {
```

Comment:
Why do we enable auth only for `https`? the `http` registry server may also require auth.

Reply:
I misunderstood, my local localhost registry does not have auth, so there will be an error because I can't get the keychain, so here I only create a RoundTripper with auth for HTTPS